.github/workflows/continuous-integration.yml

```diff
@@ -0,0 +1,266 @@
+          INI=$(echo "${CONFIG_JSON}" | jq -c "$JQ_SCRIPT")
+          echo "Using PHP INI settings: ${INI}"
+          echo "::set-output name=ini::${INI}"
+
+  unit-tests:
+    name: Unit Tests
+    runs-on: ubuntu-latest
+    needs: [config, php-setup, supported-versions-matrix, dependency-versions-matrix]
+
+    strategy:
+      fail-fast: false
+      matrix:
+        php-version: ${{ fromJson(needs.supported-versions-matrix.outputs.version) }}
+        deps-version: ${{ fromJson(needs.dependency-versions-matrix.outputs.version) }}
+        include:
+          - php-version: "8.0"
+            composer-args: --ansi --no-interaction --no-progress --prefer-dist --ignore-platform-req=php
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Setup PHP
+        uses: shivammathur/setup-php@v2
+        with:
+          php-version: ${{ matrix.php-version }}
+          extensions: ${{ join(fromJSON(needs.php-setup.outputs.extensions), ', ') }}
+          ini-values: ${{ join(fromJSON(needs.php-setup.outputs.ini), ', ') }}
+          coverage: none
+          tools: composer:v2, cs2pr
+
+      - name: "Cache dependencies"
+        uses: "actions/cache@v2"
+        with:
+          path: |
+            ~/.composer/cache
+            vendor
+          key: "php-${{ matrix.php-version }}-${{ matrix.deps-version }}-${{ hashFiles('**/composer.lock') || github.ref }}"
+          restore-keys: "php-${{ matrix.php-version }}-${{ matrix.deps-version }}-${{ hashFiles('**/composer.lock') || github.ref }}"
+
+      - name: "Install dependencies (lowest)"
+        if: matrix.deps-version == 'lowest'
+        env:
+          COMPOSER_ARGS: ${{ matrix.composer-args }}
+          COMPOSER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          composer update ${COMPOSER_ARGS:-$COMPOSER_ARGS_DEFAULT} --prefer-lowest
+          composer show --ansi
+
+      - name: "Install dependencies (locked)"
+        if: matrix.deps-version == 'locked'
+        env:
+          COMPOSER_ARGS: ${{ matrix.composer-args }}
+          COMPOSER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          composer install ${COMPOSER_ARGS:-$COMPOSER_ARGS_DEFAULT}
+          composer show --ansi
+
+      - name: "Install dependencies (latest)"
+        if: matrix.deps-version == 'latest'
+        env:
+          COMPOSER_ARGS: ${{ matrix.composer-args }}
+          COMPOSER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          composer update ${COMPOSER_ARGS:-$COMPOSER_ARGS_DEFAULT}
+          composer show --ansi
+
+      - name: Setup problem matcher for PHPUnit
+        run: echo "::add-matcher${{ runner.tool_cache }}/phpunit.json"
+
+      - name: Run unit tests on PHP ${{ matrix.php-version }}
+        run: "composer test"
+
+  checks:
+    name: Checks
+    runs-on: ubuntu-latest
+    needs: [config, php-setup, supported-checks-matrix, stable-php]
+
+    strategy:
+      fail-fast: false
```

Comment:
Noticed a few of these `fail-fast`, but unclear on their strategic position :thinking:

Reply:
I only disable it on steps that use matrices and/or when I want to ensure other steps run even if there is a failure.

The two cases it's used in this workflow are with the unit-tests and the checks steps. In the unit-tests case:

- We want to run _all_ jobs in the matrix; fail fast would prevent completion of jobs after the first that failed.
- We want to run other checks even if a job in the unit-tests matrix fails.

For the checks case, we want to run all the checks, regardless of whether one fails or not.